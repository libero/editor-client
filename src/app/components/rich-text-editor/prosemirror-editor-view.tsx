import React from 'react';
import { EditorState, Transaction } from 'prosemirror-state';
import { EditorProps, EditorView } from 'prosemirror-view';
import classnames from 'classnames';
import 'prosemirror-view/style/prosemirror.css';

import './prosemirror-styles.scss';

interface ProseMirrorEditorViewState {
  editorState: EditorState;
}

export interface ProseMirrorEditorViewProps {
  className?: string;
  editorState: EditorState;
  options?: Partial<EditorProps>;
  onChange: (tx: Transaction) => void;
}

export class ProseMirrorEditorView extends React.Component<ProseMirrorEditorViewProps, ProseMirrorEditorViewState> {
  public props;
  public editorView: EditorView;

  focus() {
    // debugger;
    this.editorView.focus();
  }

  blur() {
    (this.editorView.dom as HTMLDivElement).blur();
  }

  shouldComponentUpdate(nextProps) {
    if (nextProps.editorState !== this.editorView.state) {
      this.editorView.updateState(nextProps.editorState);
    }
    return false;
  }

  componentWillUnmount() {
    if (this.editorView) {
      this.editorView.destroy();
    }
  }

  render() {
    // Render just an empty div which is then used as a container for an
    // EditorView instance.
    return <div ref={this.createEditorView} className={classnames('prosemirrorContainer', this.props.className)} />;
  }

  private createEditorView = (element: HTMLElement) => {
    if (element) {
      const additionalOptions = this.props.options || {};
      this.editorView = new EditorView(element, {
        ...additionalOptions,
        state: this.props.editorState,
        dispatchTransaction: (tx: Transaction) => this.props.onChange(tx)
      });
    }
  };
}
