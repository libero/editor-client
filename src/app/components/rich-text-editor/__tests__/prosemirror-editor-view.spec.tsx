import React from 'react';
import { create } from 'react-test-renderer';
import { EditorState } from 'prosemirror-state';
import { ProseMirrorEditorView } from '../prosemirror-editor-view';
import { EditorView } from 'prosemirror-view';
import { mount } from 'enzyme';

jest.mock('prosemirror-view');

const createNodeMock = (element) => {
  return element;
};

describe('prosemirror view', () => {
  let mockEditorView;

  beforeEach(() => {
    mockEditorView = {
      updateState: jest.fn(),
      destroy: jest.fn()
    };
    (EditorView as jest.Mock).mockImplementation(() => mockEditorView);
  });

  it('should render', () => {
    const sampleState = new EditorState();
    const onChangeHandler = jest.fn();

    const component = create(<ProseMirrorEditorView editorState={sampleState} onChange={onChangeHandler} />, {
      createNodeMock
    });
    expect(component).toMatchSnapshot();
    expect(EditorView).toBeCalledWith(
      { props: {}, type: 'div' },
      { state: sampleState, dispatchTransaction: expect.any(Function) }
    );
  });

  it('should trigger onChange', () => {
    const sampleState = new EditorState();
    const onChangeHandler = jest.fn();

    create(<ProseMirrorEditorView editorState={sampleState} onChange={onChangeHandler} />, { createNodeMock });
    const { dispatchTransaction } = (EditorView as jest.Mock).mock.calls[0][1];

    const tx = {};
    dispatchTransaction(tx);
    expect(onChangeHandler).toHaveBeenCalledWith(tx);
  });

  it('should update EditorView when props change', () => {
    const sampleState = new EditorState();
    const onChangeHandler = jest.fn();

    const component = mount(<ProseMirrorEditorView editorState={sampleState} onChange={onChangeHandler} />);
    const newState = new EditorState();
    component.setProps({ onChange: onChangeHandler, editorState: newState });
    expect(mockEditorView.updateState).toBeCalledWith(newState);
  });

  it('should destroy EditorView when unmount', () => {
    const sampleState = new EditorState();
    const onChangeHandler = jest.fn();

    const component = mount(<ProseMirrorEditorView editorState={sampleState} onChange={onChangeHandler} />);
    component.unmount();
    expect(mockEditorView.destroy).toBeCalled();
  });

  afterEach(() => {
    (EditorView as jest.Mock).mockReset();
  });
});
