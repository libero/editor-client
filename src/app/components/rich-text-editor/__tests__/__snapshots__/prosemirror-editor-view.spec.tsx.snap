// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prosemirror view should render 1`] = `<div />`;
